filters/tags_filter.go

```diff
@@ -0,0 +1,159 @@
+)
+
+// SortedTagIterator iterates over a set of tag names and values
+// sorted by tag names in ascending order
+type SortedTagIterator interface {
+	// Next returns true if there are more tag names and values
+	Next() bool
+
+	// Current returns the current tag name and value
+	Current() (string, string)
+
+	// Err returns any errors encountered
+	Err() error
+
+	// Close closes the iterator
+	Close()
+}
+
+// NewSortedTagIteratorFn creates a tag iterator given an id
+type NewSortedTagIteratorFn func(id string) SortedTagIterator
+
+// tagFilter is a filter associated with a given tag
+type tagFilter struct {
+	name        string
+	valueFilter Filter
+}
+
+func (f *tagFilter) String() string {
+	return fmt.Sprintf("%s:%s", f.name, f.valueFilter.String())
+}
+
+type tagFiltersByNameAsc []tagFilter
+
+func (tn tagFiltersByNameAsc) Len() int           { return len(tn) }
+func (tn tagFiltersByNameAsc) Swap(i, j int)      { tn[i], tn[j] = tn[j], tn[i] }
+func (tn tagFiltersByNameAsc) Less(i, j int) bool { return tn[i].name < tn[j].name }
+
+// tagsFilter contains a list of tag filters.
+type tagsFilter struct {
+	filters []tagFilter
+	iterFn  NewSortedTagIteratorFn
+	op      LogicalOp
+}
+
+// NewTagsFilter create a new tags filter
+func NewTagsFilter(tagFilters map[string]string, iterFn NewSortedTagIteratorFn, op LogicalOp) (Filter, error) {
+	filters := make([]tagFilter, 0, len(tagFilters))
+	for name, value := range tagFilters {
+		valFilter, err := NewFilter(value)
+		if err != nil {
+			return nil, err
+		}
+
+		filters = append(filters, tagFilter{
+			name:        name,
+			valueFilter: valFilter,
+		})
+	}
+	sort.Sort(tagFiltersByNameAsc(filters))
+	return &tagsFilter{
+		filters: filters,
+		iterFn:  iterFn,
+		op:      op,
+	}, nil
+}
+
+func (f *tagsFilter) String() string {
+	separator := " " + string(f.op) + " "
+	var buf bytes.Buffer
+	numFilters := len(f.filters)
+	for i := 0; i < numFilters; i++ {
+		buf.WriteString(f.filters[i].String())
+		if i < numFilters-1 {
+			buf.WriteString(separator)
+		}
+	}
+	return buf.String()
+}
+
+func (f *tagsFilter) Matches(id string) bool {
+	if len(f.filters) == 0 {
+		return true
+	}
+	iter := f.iterFn(id)
+	defer iter.Close()
+
+	currIdx := 0
+	for iter.Next() && currIdx < len(f.filters) {
+		name, value := iter.Current()
+		if name < f.filters[currIdx].name {
+			continue
+		}
+
+		if name > f.filters[currIdx].name {
+			if f.op == Conjunction {
+				// For AND, if the current filter tag doesn't exist, bail immediately
+				return false
+			}
+
+			// Iterate filters for the OR case
+			currIdx++
+			for currIdx < len(f.filters) && name > f.filters[currIdx].name {
+				currIdx++
+			}
+
+			if currIdx == len(f.filters) {
+				// Past all filters
+				return false
+			}
+		}
+
+		match := f.filters[currIdx].valueFilter.Matches(value)
+		if match && f.op == Disjunction {
+			return true
+		}
+
+		if !match && f.op == Conjunction {
+			return false
+		}
+
+		currIdx++
+	}
+
+	if iter.Err() != nil {
```

Comment:
nit: could simplify these to `return iter.Err() == nil && f.op == Conjunction && currIdx == len(f.filters)`